### PR TITLE
Refactor BuildNodeIdForType such that it only builds nodes,

### LINKED ISSUE #2967: //kythe/docs/schema:schema build broken
bazel build //kythe/docs/schema returns errors like:

```
kythe/java/com/google/devtools/kythe/analyzers/java/indexer: Cannot locate runfiles directory. (Set $JAVA_RUNFILES to inhibit searching.)
[ FAILED INDEX: Classes can be annotated. ]
```

### PR DESCRIPTION
**Pull request by shahms**
emitting semantic nodes as necessary, but the actual decoration of types
and ranges is done via TraverseTypeDecl.

Squashed commit of the following:

commit 7f91c77f70a872e3f04056bfedbd817f1184a01b
Merge: b91b2b1b 24d864d1
Author: Shahms King <shahms@google.com>
Date:   Tue Sep 25 09:11:04 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit b91b2b1beb27e047903ca23a94250f95b2f44204
Author: Shahms King <shahms@google.com>
Date:   Tue Sep 25 09:10:45 2018 -0700

    Include comment on the .ForReference accessor

commit bd5bdd5b8d69e7982d8ba76897f23eaabe63023e
Author: Shahms King <shahms@google.com>
Date:   Tue Sep 25 09:08:59 2018 -0700

    Finish moving to TypeLoc traversal-based node emission, re-enable duplicate checks in tests.

commit 24d864d19aa2091998347fe06cd05e8e35e2a432
Author: Daniel Moy <danielmoy@google.com>
Date:   Mon Sep 24 16:40:55 2018 -0700

    Use filepath.Dir & filepath.Base instead of filepath.Split. (#3053)

    Specifically to avoid #3051 we need to make sure we get a "." for the
    local directory instead of "", because `ioutil.TempFile` will helpfully
    prepend /tmp if you pass "" to dir.

commit 50a75f83534f9f89465bb3389b39afcbfbcb698b
Author: Daniel Moy <danielmoy@google.com>
Date:   Mon Sep 24 16:17:45 2018 -0700

    Use runextractor in docker extraction bits, not the old scripts (#3052)

    Also removes MvnPomPreprocessor, mvn-extract.sh

commit 334a36e3076792895fce132a914ee3cd905a2156
Author: Sal <salguarnieri@google.com>
Date:   Mon Sep 24 15:02:39 2018 -0700

    Add helper method for Kotlin indexer (#3050)

    The Kotlin indexer uses this helper method.

commit 18b736459fa52b42901b958112c7fda2ddad2f83
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 24 14:48:43 2018 -0700

    Migrate ObjCObject

commit 6dd322c632b8cda10b363ba9fb84e70e81332cfc
Author: M. J. Fromberger <michael.j.fromberger@gmail.com>
Date:   Mon Sep 24 13:51:47 2018 -0700

    Bazel extractor: Remove a redundant call to config.FixUnit. (#3048)

    The underlying extraction machinery already invokes the fixup, so it is not
    necessary to repeat it when extracting to a file. It is harmess apart from
    being wasteful of resources to do so as long as the fixup is idempotent -- but
    since that is not required, so don't (unintentionally) rely on it.

commit a9c1a237a19c9eaec7067fffda5427e5a404492d
Author: Daniel Moy <danielmoy@google.com>
Date:   Mon Sep 24 13:04:54 2018 -0700

    Add runextractor to kythe-java-extractor-artifacts image. (#3046)

    * Add runextractor to kythe-java-extractor-artifacts image.

    Soon we can ax mvn-extract.sh, hopefully.

commit defa3f2a17988467c6c181c9709571decc5ac04d
Author: Cody Schroeder <schroederc@google.com>
Date:   Mon Sep 24 11:41:49 2018 -0700

    Remove platform directory prefixes from go_binary output paths (#3047)

commit 141f590cf549f90021927d145dae105ac51edf32
Author: Cody Schroeder <schroederc@google.com>
Date:   Mon Sep 24 11:34:55 2018 -0700

    Add //kythe/docs/schema to Buildbot release tests (#3045)

    The other asciidoc_with_verifier targets have their 'manual' tag removed so they are tested in all configurations.

    The bazel-opt builder has been folded into the bazel-release builder.

commit f2669d7991c1a90a7f08256b918df5591a90f2d0
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 24 10:53:56 2018 -0700

    Fix typo preventing VisitObjCTypeParamTypeLoc from being called

commit ca6ee744c2eac7d677ee269c9d05167e07f76235
Author: Cody Schroeder <schroederc@google.com>
Date:   Mon Sep 24 10:20:50 2018 -0700

    Ensure asciidoc invocation gets data runfiles (for Java) (#3044)

    Fixes #2967

commit 9085728bf56079d8c1f7b7e5cea11493dc6e98bb
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 24 10:15:53 2018 -0700

    Migrate Attribute and DependentAddressSpace

commit d0344158c918d5febd24f765acb7f960398be2fe
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 24 09:59:05 2018 -0700

    minor incremental cleansup for ObjCObject

commit 2ab7a5fb80c7a220ee0344b80dfd944296f054ea
Author: M. J. Fromberger <michael.j.fromberger@gmail.com>
Date:   Mon Sep 24 08:54:48 2018 -0700

    Bazel extractors: Emit .kzip files by default. (#3043)

    As part of #2867, switch the default format emitted by the Go-based Bazel
    extractors to be .kzip instead of .kindex. This means the name of the tool is
    now no longer accurate, but we can clean that up in a separate pass.

commit 4710352f7b44f361eaf3b40a061a2366aba70780
Author: zrlk <zrlk@users.noreply.github.com>
Date:   Fri Sep 21 15:17:40 2018 -0700

    Split type MarkedSource nodes containing decl names (#3041)

    We had been doing this for functions, but we should
    also do it for variables and fields.

    Addresses #3039.

commit b53f2ba16d2a1bbacdef4f0695ee6fb8b8855eba
Author: Shahms King <shahms@google.com>
Date:   Fri Sep 21 14:14:51 2018 -0700

    Migrate TemplateSpecialization

commit 1df3016ebf1269363fd08c689aa97158eb69e0f4
Author: Cody Schroeder <schroederc@google.com>
Date:   Fri Sep 21 13:02:05 2018 -0700

    Use RBE for all Bazel 0.17.1 Builtbot configurations (#3040)

commit 82cb2bdbcb66dbfb3816d4d6c855e452b3a10805
Author: Cody Schroeder <schroederc@google.com>
Date:   Fri Sep 21 12:13:16 2018 -0700

    Fix java_verifier_tests in RBE environment (#3038)

commit 2ddcbde9dc75f6b1ce36d21bcc6c7a835ac12eb7
Author: Cody Schroeder <schroederc@google.com>
Date:   Fri Sep 21 10:18:54 2018 -0700

    Fork Flogger to build Java targets in RBE environment (#3037)

    This is a patch until https://github.com/google/flogger/pull/37 is closed.

commit 0546b3e6813a661e979a2e8e1400f07a47286ec7
Author: Cody Schroeder <schroederc@google.com>
Date:   Fri Sep 21 09:48:47 2018 -0700

    Fix asciidoc targets in RBE environment (#3036)

commit 9ba3472fddbc1a284cc6cb7d21064648e06906ab
Author: Cody Schroeder <schroederc@google.com>
Date:   Fri Sep 21 09:30:28 2018 -0700

    Fix verifier tests in RBE environment (#3035)

commit ec6f8afeedff5c7ac6c3f4925807065aeb2e1209
Merge: 434903a3 2b4f09f0
Author: Shahms King <shahms@google.com>
Date:   Fri Sep 21 09:08:06 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit 434903a3f6fa58bc7d0f2c992003fea1075d9460
Author: Shahms King <shahms@google.com>
Date:   Wed Sep 19 14:25:19 2018 -0700

    Add NodeSet::AsOptional and use it to migrate InjectedClassName logic

commit 522d11a633ed69d279ad00f8d4ac3c2590ae0bfd
Author: Shahms King <shahms@google.com>
Date:   Wed Sep 19 14:00:27 2018 -0700

    Add a NodeSet class representing primary and secondary nodes, migrate Record using this

commit da2f5bf8286f573583fedb127cb8ab4bf9b0d39e
Author: Shahms King <shahms@google.com>
Date:   Wed Sep 19 09:42:35 2018 -0700

    Remove now-unnecessary code and rename BuildNodeIdFor*TypeLoc to remove TypeLoc

commit a78856fcff81b50b1f5318fd7deafd7daf34eef6
Author: Shahms King <shahms@google.com>
Date:   Tue Sep 18 11:05:42 2018 -0700

    Refactor Typedef, Elaborated and move tnominal construction to a helper.

commit a38c2f0b740e7814886b8ffd8258dd994cba3d70
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 13:47:14 2018 -0700

    Refactor Paren, Decltype and FunctionNoProto

commit c5469ea25bce4b4b8af46d66dd65ad7a3373a136
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 13:23:23 2018 -0700

    Actualy use refactored DependentSizedArray

commit 0ed74d00af39ae0e191d25b699ce85eb39363b96
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 11:19:57 2018 -0700

    Refactor DependentSizedArrayTypeLoc

commit 4803cb28591f8dab7bd1e353771f84bb866479dd
Merge: d6658a41 d2481c38
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 10:11:49 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit d6658a414b1d44224d1a35caff74b485ac779068
Merge: 54ce7876 30e66fc6
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 10:06:00 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit 54ce7876518259b17aafd33e02955ebbb58e0390
Author: Shahms King <shahms@google.com>
Date:   Mon Sep 17 10:05:48 2018 -0700

    Scope the VarDecl TSI type override; refactor ConstantArray and IncompleteArray

commit 11ea6edca3c135845c7506b7e1251e73af3102fb
Author: Shahms King <shahms@google.com>
Date:   Fri Sep 14 14:40:53 2018 -0700

    refactor Pointer, Reference, Deduced and Qualified types. Include a hack fixing Deduced types.

commit a7e6d9510d1148cd40672b8fa7851e8ccfa734b4
Merge: 0d0d6ece 3ccef495
Author: Shahms King <shahms@google.com>
Date:   Thu Sep 13 12:46:57 2018 -0700

    Merge branch 'objc-simplify' into type-loc

commit 3ccef4957cd97cd8b1417631c4d5ce6115d58b36
Author: Shahms King <shahms@google.com>
Date:   Thu Sep 13 12:46:12 2018 -0700

    Further simplify eligible BuildNodeIdForType calls

commit d740dd949733c09fe83746aa663710ab5d36d99e
Author: Shahms King <shahms@google.com>
Date:   Thu Sep 13 09:11:31 2018 -0700

    Cleanup remaining unnecessary cluttered BuildNodeIdForType calls

commit 8b906b1cd98fc81575876b6bc176c546642ff232
Author: Shahms King <shahms@google.com>
Date:   Thu Sep 13 09:02:14 2018 -0700

    Simplify relevant calls to BuildNodeIdForType

commit 0d0d6ece7023474476505e3d2303aafdb4969830
Author: Shahms King <shahms@google.com>
Date:   Thu Sep 13 08:30:26 2018 -0700

    Prepare for additional refactorings

commit 21d4f347578dfc645ea4c1d5d14e1e3532a89e58
Merge: e0ed9598 a83cfd6d
Author: Shahms King <shahms@google.com>
Date:   Wed Sep 12 13:46:27 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit e0ed9598de622946a3454afb540d9df40af7d26a
Merge: 7f4dfede 65336ac3
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 28 10:48:47 2018 -0700

    Merge branch 'type-loc' of github.com:shahms/kythe into type-loc

commit 7f4dfede12387a775be741d83aceb97aa69934ff
Merge: 7cc6d4e3 f32f399c
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 28 10:45:18 2018 -0700

    Merge branch 'master' of github.com:google/kythe into type-loc

commit 7cc6d4e36e413300fcd412c096af0e85d63c0161
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 28 10:44:02 2018 -0700

    Actually implement BuildNodeIdForRecordTypeLoc for templates, rather than defering

commit 04c8ab863fb3da982a6bede88dd6c3ec380d87a6
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 14 15:09:51 2018 -0700

    Delegate non-template parts of BuildNodeIdForType to BuildNodeIdForRecordTypeLoc

commit 4fdcd67879ffba61a69f8b41d9b71ccce6d8a929
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 14 13:02:00 2018 -0700

    Refactor non-template RecordTypeLoc's to use new facilities

commit a973dfc646d5dcfc9342b4083fe2ad8d73b85e0d
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 14 10:55:22 2018 -0700

    Migrate ObjCInterfaceTypeLoc handling to VisitObjCInterfaceTypeLoc

commit df562e88e10f004f6f0c24cb1658850934b47958
Merge: 895b5db7 87e698d6
Author: Shahms King <shahms@google.com>
Date:   Tue Aug 14 09:08:46 2018 -0700

    Merge with master

commit 895b5db771170cf3ddb2d4e26174eb182acc5658
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 16 14:05:39 2018 -0700

    Constify some relevant member functions

commit d0b29fa5972ae1d78299573b358265f04e6f5af6
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 9 13:49:51 2018 -0800

    Correctly index qualified or nested calls to C++ constructors in more cases

    Summary: Correctly index qualified or nested calls to C++ constructors in more cases

    Reviewers: zarko

    Subscribers: #core_team

    Differential Revision: https://phabricator-dot-kythe-repo.appspot.com/D2215

commit 26d25acca4061d93edea7ec87dbd029835d803bd
Author: Shahms King <shahms@google.com>
Date:   Thu Mar 8 10:44:34 2018 -0800

    Update calls to recordTypeSpellingLocation

commit 112fb13b488859e43b262da3e7766c53510b2701
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 26 14:44:15 2018 -0800

    Remove unnecessary post-merge use of VisitNestedNamespaceSpecifierLoc

commit 3b903697094326b7e1d7d8c6a389acef7ea3e8c8
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 19 10:26:15 2018 -0800

    Move recordNominalTypeNode for EnumTypeLoc to Visit from Build

commit 61c6e68a5a1b7956c8bec73a7e1bdb953fe512bb
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 19 09:24:49 2018 -0800

    Allow BuildNodeIdForTemplateTypeParmTypeLoc to fail

commit 7a31f2daf97b150455405fde81f240bb85ea0507
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 15:12:54 2018 -0800

    Split/move BuildNodeIdForType(TemplateTypeParmTypeLoc) to its own function

commit 36b11610020b54c4cd91a6fb3dbf885acc3affd5
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 14:10:30 2018 -0800

    Pull Enum and Builtin into their own BuildNodeIdFor* member functions

commit d499ac91d5db9b466c9e9d69a45cb74551f30851
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 13:39:32 2018 -0800

    prefer assignment and remove optional from ApplyBuiltinTypeConstructor

commit 4978eb27282d0c14f28ba02e2f899e3ab5e1063a
Author: Shahms King <shahms@google.com>
Date:   Wed Jan 17 08:20:37 2018 -0800

    Use an immediate lambda for traversing dependent specializations

commit c0d4015f6f5233864f12f9a2d3280eeb45676b22
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 16:06:35 2018 -0800

    Migrate to VisitSubstTemplateTypeParmTypeLoc

commit 215fb9cb6d274f9c38d7a083b923d248dbeea67b
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 15:57:37 2018 -0800

    Move graph emission into VisitTemplateTypeParmTypeLoc from BuildNodeIdForType

commit 3ed8e5d9a36adfb13c02a962fa8e1d61442f5e6f
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 15:13:15 2018 -0800

    Update todo messages to be more accurate

commit c743c1ef9b00b8b53a12503a90e9b22f3427ca05
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 14:45:19 2018 -0800

    Move anchor/ref emission for enum types to VisitEnumTypeLoc rather than BuildNodeIdForType. Add test for forward declared enums.

commit 7a204d86aedba0cdb5691525513eb8dc173f070f
Author: Shahms King <shahms@google.com>
Date:   Sat Jan 13 16:49:39 2018 -0800

    WIP - Move Builtin ref emission from BuildNodeIdForType to VisitBuiltinTypeLoc

commit 4cb232b66c79dafc21d63a925c028e7568d1fb8e
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 12 08:21:57 2018 -0800

    WIP - Remove VisitNestedNameSpecifier in favor of direct Visit*TypeLoc

    Note: ignore_dups is changed to `True` in the tests and
    the TypeLoc::TemplateTypeParm case in BuildNodeIdForType
    has been modified. Both of these should be reverted before
    merging.

commit 65336ac39fb61304cf075a5736fd94d90c10b6d2
Merge: 3b498be3 043421a6
Author: Shahms King <shahms@google.com>
Date:   Wed Mar 21 15:20:34 2018 -0700

    Merge branch 'master' into type-loc

commit 3b498be304b354186fbc3d6f5f8a608acdde5b63
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 16 14:05:39 2018 -0700

    Constify some relevant member functions

commit f694aaed6ef44129830e375f8be6acc8e8a37083
Merge: e9646a5a 0287c94c
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 16 08:51:20 2018 -0700

    Merge branch 'master' into type-loc

commit e9646a5a0be74793419e2aecfa1234a807e344f1
Merge: 6f45a00d 0a3afc9b
Author: Shahms King <shahms@google.com>
Date:   Thu Mar 15 10:53:58 2018 -0700

    Merge branch 'master' into type-loc

commit 6f45a00dd35fa0d9f81fe6226c8493427836db23
Merge: e903193a 04b9e4eb
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 9 14:01:02 2018 -0800

    Merge branch 'qual-ctor' into type-loc

commit 04b9e4ebd251aa29579fec011c7272e2bf1bcdf9
Author: Shahms King <shahms@google.com>
Date:   Fri Mar 9 13:49:51 2018 -0800

    Correctly index qualified or nested calls to C++ constructors in more cases

    Summary: Correctly index qualified or nested calls to C++ constructors in more cases

    Reviewers: zarko

    Subscribers: #core_team

    Differential Revision: https://phabricator-dot-kythe-repo.appspot.com/D2215

commit e903193a966e9d22989bd05f295cd41d71735e0b
Author: Shahms King <shahms@google.com>
Date:   Thu Mar 8 10:44:34 2018 -0800

    Update calls to recordTypeSpellingLocation

commit 0f0dfcaaf03d2dfdcc40762b9ee2d544005875b3
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 26 14:44:15 2018 -0800

    Remove unnecessary post-merge use of VisitNestedNamespaceSpecifierLoc

commit 767270221ffe9d125192a733902c13eafd9837b7
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 19 10:26:15 2018 -0800

    Move recordNominalTypeNode for EnumTypeLoc to Visit from Build

commit 3badf8beea62d6d753bcdcd279e9626fb9c899dc
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 19 09:24:49 2018 -0800

    Allow BuildNodeIdForTemplateTypeParmTypeLoc to fail

commit 57abd2518ee43f9053d8c4f8d6aef3fc5b6fbc5a
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 15:12:54 2018 -0800

    Split/move BuildNodeIdForType(TemplateTypeParmTypeLoc) to its own function

commit 4b5df2438b4088e19a868d4c0715c95b20f83c87
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 14:10:30 2018 -0800

    Pull Enum and Builtin into their own BuildNodeIdFor* member functions

commit aa7e5009c3cb8f9e5590199c2745d2b9a7b1cb50
Author: Shahms King <shahms@google.com>
Date:   Thu Jan 18 13:39:32 2018 -0800

    prefer assignment and remove optional from ApplyBuiltinTypeConstructor

commit f200c718a8a388cd71738174f77447570d43e10c
Author: Shahms King <shahms@google.com>
Date:   Wed Jan 17 08:20:37 2018 -0800

    Use an immediate lambda for traversing dependent specializations

commit dfafebd86f785a2b1f3b05071bb3a39eb280aa48
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 16:06:35 2018 -0800

    Migrate to VisitSubstTemplateTypeParmTypeLoc

commit ceda0a40acd0593a9decc3f158ddc8f403fc87c2
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 15:57:37 2018 -0800

    Move graph emission into VisitTemplateTypeParmTypeLoc from BuildNodeIdForType

commit b6c0304112b6a5c719ba0c3c2f99dded36dc755d
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 15:13:15 2018 -0800

    Update todo messages to be more accurate

commit 4b5ca025ff9fc10cd51a1e438898c68b36fe605e
Author: Shahms King <shahms@google.com>
Date:   Sun Jan 14 14:45:19 2018 -0800

    Move anchor/ref emission for enum types to VisitEnumTypeLoc rather than BuildNodeIdForType. Add test for forward declared enums.

commit 36d02269f6b65bb12ca6576412b7b0ca0218b72f
Author: Shahms King <shahms@google.com>
Date:   Sat Jan 13 16:49:39 2018 -0800

    WIP - Move Builtin ref emission from BuildNodeIdForType to VisitBuiltinTypeLoc

commit 6d400ba2e717f8eee836f79762c586b152e15ef1
Author: Shahms King <shahms@google.com>
Date:   Fri Jan 12 08:21:57 2018 -0800

    WIP - Remove VisitNestedNameSpecifier in favor of direct Visit*TypeLoc

    Note: ignore_dups is changed to `True` in the tests and
    the TypeLoc::TemplateTypeParm case in BuildNodeIdForType
    has been modified. Both of these should be reverted before
    merging.

### CODE REVIEW

**Comment by shahms** [comment]
Apologies for the size, there wasn't really a good way of doing this incrementally :-(

**Comment by schroederc** [comment]
That squash log seems wrong (it includes me).  Can you rework the Pull Request to just include the branch with all of the commits unsquashed?

**Comment by shahms** [comment]
@schroederc  Done but ... I don't think it really helps clarify the PR at all.

**Comment by shahms** [comment]
The squash log is correct because it includes a lot of merged changes from master, some of which you committed :-)

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 1]
```diff
@@ -232,6 +235,26 @@ const clang::Decl* FindImplicitDeclForStmt(
   }
   return nullptr;
 }
+
+// Overrides the type from `TSI` with that in `Type` restoring the original upon
+// destruction.
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/cxx/IndexerASTHooks.cc]: Clearer?

File: kythe/cxx/indexer/cxx/IndexerASTHooks.cc [hunk 2]
```diff
@@ -1699,32 +1718,144 @@ void IndexerASTVisitor::AddChildOfEdgeToDeclContext(
...
+  return EmitTypeLocNodes(TL);
+}
+
+bool IndexerASTVisitor::VisitRecordTypeLoc(clang::RecordTypeLoc TL) {
+  return EmitTypeLocNodes(TL);
+}
+
+bool IndexerASTVisitor::VisitObjCInterfaceTypeLoc(
+    clang::ObjCInterfaceTypeLoc TL) {
+  // return EmitTypeLocNodes(TL);
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/cxx/IndexerASTHooks.cc]: Done, thanks.

**Comment by shahms** [comment]
Comments address in #3057

**Pull request closed without merging**